File: src/main/java/thrive/TrieUtils.java
package thrive;

class TrieUtils {
    static int mask(int index, int shift, int bits) {
        return (index >>> shift) & ((1 << bits) - 1);
    }

    static int indexS(short bitmap, short pos) {
        return Integer.bitCount(Short.toUnsignedInt(bitmap) & (Short.toUnsignedInt(pos) - 1));
    }

    static int indexI(int bitmap, int pos) {
        return Integer.bitCount(bitmap & (pos - 1));
    }

    static int indexL(long bitmap, long pos) {
        return Long.bitCount(bitmap & (pos - 1));
    }

    static void copyInto(Object src, Object dest, int destinationOffset, int startIndex, int endIndex) {
        System.arraycopy(src, startIndex, dest, destinationOffset, endIndex - startIndex);
    }
}

File: src/main/kotlin/thrive/TrieUtils.kt
package thrive

import java.util.ArrayList

internal fun mask(index: Int, shift: Int, bits: Int): Int = (index ushr shift) and ((1 shl bits) - 1)
internal fun index(bitmap: Int, pos: Int): Int = Integer.bitCount(bitmap and (pos - 1))

fun <T> ArrayList<T>.pop(): T = this.removeAt(this.size - 1)
fun <T> ArrayList<T>.peek(): T = this[this.size - 1]